artemis-commons/src/main/java/org/apache/activemq/artemis/utils/collections/BucketMap.java

```diff
@@ -0,0 +1,177 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements. See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License. You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.activemq.artemis.utils.collections;
+
+import io.netty.util.collection.IntObjectHashMap;
+import io.netty.util.collection.IntObjectMap;
+import java.util.Collection;
+import java.util.Map;
+import java.util.Set;
+import java.util.function.BiConsumer;
+import java.util.function.BiFunction;
+import java.util.function.Function;
+
+/**
+ * BucketMap, stores values against a bucket, where the bucket used is based on the provided key objects hash.
+ *
+ * As such where keys compute to the same bucket they will act on that stored value, not the unique specific keys as you'd normally expect from a map.
+ *
+ * The number of buckets is provided at construction.
+ *
+ * Its initial use, is in QueueImpl, where we want to bucket message groups to consumers.
+ *
+ * @param <K> the key type.
+ * @param <V> the value type.
+ */
+public class BucketMap<K, V> implements Map<K, V> {
+
+   private final IntObjectMap<V> buckets = new IntObjectHashMap<>();
```

Comment:
I'm not sure if I have already left this comment but % is really an heavy operation to be done on hot path so I suggest to use the power of 2 + mask trick

Reply:
And..idea..given that you are % on bucket count is you could have collisions. Maybe you don't need a map here but just an array bucketCount sized...

Reply:
Collisions are 100% expected we are bucketing. E.g if bucket size is two and the key was integer with keys 1 and 3 i actually expect them to collide and goto the same bucket.

Re map vs array, i was toying on the same thing, downside with array is up front will need to allocate the memory for that array even if not all buckets used. Benefit of using underlying map is it will grow as needed.

Reply:
Not sure...consider that the map will allocate both the int and obj arrays multiplied by the load factor. And we will pay the rehashing too..a fixed size thing will be more "direct" (less pointer chasing) and simple (just obj array, never resized). 
Is just a personal opinion eh, I'm anyway happy about it.

Reply:
I will have some time on the next week and I will look better into this. ATM I have just my phone :)

Reply:
Ive actually found a small issue in Netty's implementation...they don't support values().itertator().remove(), annoyingly theres no reason either, the underlying iterator they facade would support it .... grrrrr

Reply:
@franz1981 issue with array is imaging someone on their broker (that someone being me) wants to ensure group buckets by default across the broker to say 20k, as such sets address setting match = #.

Then for every queue will create a 20k array, even if the producer on the address doesnt set message groups.

Reply:
I could implement it on Netty bud, np :)

Reply:
@franz1981 just because of that netty thing, i may have to have our own version again.

What im thinking of doing, is so we dont hold back this feature and it makes the 2.7.0 cut, is clone in the netty varient with a fix to our collections (they're apache license so no license issues in doing this, ill just add an attribution to the java doc), and then if netty fix it, then we can remove it again at some fututre point.